File: tests/src/commonTest/kotlin/primitives/CallbackArgPrimitive.kt
package primitives

import withLib
import kotlin.test.Test
import kotlin.test.assertTrue

class CallbackArgPrimitive {

    @Test
    fun callbackVoid() = withLib {
        var pass = false
        natives.test.callbackVoid { pass = true }
        assertTrue(pass)
    }

    @Test
    fun callbackArgChar() = withLib {
        assertTrue(natives.test.callbackArgChar {
            println("${it.code} | ${it == 'a'}")
            it == 'a'
        })
    }

    @Test
    fun callbackArgBoolean() = withLib {
        assertTrue(natives.test.callbackArgBoolean { it })
    }

    @Test
    fun callbackArgByte() = withLib {
        assertTrue(natives.test.callbackArgByte { it == 1.toByte() })
    }

    @Test
    fun callbackArgShort() = withLib {
        assertTrue(natives.test.callbackArgShort { it == 1.toShort() })
    }

    @Test
    fun callbackArgInt() = withLib {
        assertTrue(natives.test.callbackArgInt { it == 1 })
    }

    @Test
    fun callbackArgLong() = withLib {
        assertTrue(natives.test.callbackArgLong { it == 1.toLong() })
    }

    @Test
    fun callbackArgFloat() = withLib {
        assertTrue(natives.test.callbackArgFloat { it == 1.1f })
    }

    @Test
    fun callbackArgDouble() = withLib {
        assertTrue(natives.test.callbackArgDouble { it == 1.1 })
    }

    @Test
    fun callbackArgString() = withLib {
        assertTrue(natives.test.callbackArgString { it == "test string" })
    }

    @Test
    fun callbackArgCallback() = withLib {
        val toPass = {}
        assertTrue(natives.test.callbackArgCallback(toPass) { it == toPass })
    }
}

File: tests/natives/test/src/primitives/callback_arg_primitive.c
#include <api.h>

void callbackVoid(VoidCallback* arg) {
    arg->invoke(arg);
}

KBoolean callbackArgChar(CallbackPassChar* arg) {
    return arg->invoke(arg, 'a');
}

KBoolean callbackArgBoolean(CallbackPassBoolean* arg) {
    return arg->invoke(arg, true);
}

KBoolean callbackArgByte(CallbackPassByte* arg) {
    return arg->invoke(arg, 1);
}

KBoolean callbackArgShort(CallbackPassShort* arg) {
    return arg->invoke(arg, 1);
}

KBoolean callbackArgInt(CallbackPassInt* arg) {
    return arg->invoke(arg, 1);
}

KBoolean callbackArgLong(CallbackPassLong* arg) {
    return arg->invoke(arg, 1);
}

KBoolean callbackArgFloat(CallbackPassFloat* arg) {
    return arg->invoke(arg, 1.1f);
}

KBoolean callbackArgDouble(CallbackPassDouble* arg) {
    return arg->invoke(arg, 1.1);
}

KBoolean callbackArgString(CallbackPassString* arg) {
    return arg->invoke(arg, makeKString("test string", 11));
}

KBoolean callbackArgCallback(VoidCallback* pass, CallbackPassCallback* arg) {
    return arg->invoke(arg, pass);
}